src/Download.jl

```diff
@@ -0,0 +1,230 @@
+        rm(path, force=true)
+        error("Download $url failed, status code $(response.status)")
+    end
+    if file_hash !== nothing
+        calc_hash = hash_file(path)
+        if calc_hash != file_hash
+            msg  = "File hash mismatch!\n"
+            msg *= "  Expected SHA2-256: $file_hash\n"
+            msg *= "  Computed SHA2-256: $calc_hash"
+            rm(path)
+            error(msg)
+        end
+    end
+    return path
+end
+
+"""
+    unpack(tarball, [ path ]; [ tree_hash = <sha1> ]) -> path
+
+Extract `tarball` to `path`, checking that the resulting tree has the git tree
+SHA1 hash provided (if any is). All arguments are strings. When the tree is
+unpacked, the file `\$path/.tree_info.toml` is written containing these entries:
+
+* `git-tree-sha1`: the git tree SHA1 hash of the extracted tarball contents
+* `contents`: a table of extracted paths mapped to each file's type, one of
+  * `directory` — a directory
+  * `file` — a plain file, not executable
+  * `executable` — a file with the user executable bit set
+  * `symlink: \$link` — a symbolic link followed by the link target
+
+If `path` is already a tree and `tree_hash` is passed, the tree is checked:
+
+* If `\$path/.tree_info.toml` exists and is a valid TOML file with a top-level
+  `git-tree-sha1` key, then the associated value is compared to `tree_hash` and
+  if they match the tree is presumed to have that hash and nothing is done.
+
+* Otherwise, the existing tree at `path` is removed and replaced by the result
+  of extracting `path` where `\$path/.tree_info.toml` is written as described.
+
+If `tree_hash` is provided and after extracting the given tarball its computed
+tree hash does not match the given hash value, `path` is deleted and an error
+is thrown indicating the hash mismatch. Thus, successful completion guarantees
+that the unpacked tree has the correct content.
+
+If `path` is not provided, `path = tempname()` is used and returned.
+"""
+function unpack(
+    tarball :: AbstractString,
+    path :: AbstractString = tempname();
+    tree_hash :: Union{AbstractString, Nothing} = nothing,
+)
+   unpack_core(() -> tarball, path, tree_hash = tree_hash)
+end
+
+function unpack_core(
+    get_tarball :: Function,
+    path :: AbstractString;
+    tree_hash :: Union{AbstractString, Nothing},
+)
+    tree_info = joinpath(path, ".tree_info.toml")
+    tree_hash = normalize_tree_hash(tree_hash)
+    if tree_hash === nothing || isdir(path)
+        if tree_hash !== nothing && isfile(tree_info)
+            tree_info_hash = try
+                get(TOML.parsefile(tree_info), "git-tree-sha1", nothing)
+            catch err
+                @warn "invalid TOML" path=tree_info err
+            end
+            tree_info_hash == tree_hash && return path
+        end
+        rm(path, force=true, recursive=true)
+    end
+    tarball = get_tarball()
+    contents = Dict{String,String}()
+    symlinks = Dict{String,String}()
+    open(`gzcat $tarball`) do io
```

Comment:
So this is hardcoded to only support gzipped files. Is the plan to extend this to use `CodecBzip2.jl` etc. to support more formats? Like, have a table which maps various file extensions to different transformations?

Is there also a `Zip.jl` which can deal with .zip files?

Reply:
I don't want to add tons of Julia dependencies here because this is going to get baked into the Julia sysimg, so the plan is TBD. I may try to make this portable using JLLs, which I'm less concerned about baking in since they're not in the sysimg and not evolving fast. I may also use gzip to bootstrap installing support for other compression formats.